### ISSUE TITLE
"Last week", "First week" and nth week of the month (as opposed to year)

### ISSUE THREAD

**Issue by d1g** [opening-hours/opening_hours_grammar#4]
Users were [requesting support here](http://www.openstreetmap.org/way/120648442). It is impossible to compute because: 
- Ambiguity in the definition
  - fractional weeks
    - (case 1) a week that starts with "start of the week" day (including incomplete 1-6 day weeks)
  - complete weeks
    - (case 2) a week that starts with "start of the week" day (including incomplete 1-6 day weeks) + missing days from the next month (in case of "last week")
    - (case 3) a week that contains all weekdays once (start and end date varies from year to year) and doesn't "wrap" in calendar presentation (country-dependant)
    - (case 4) an interval with last 7 days (last day of the month -7days)
- Absolutely unclear semantic when "start of the week" is used (country-dependent); if a value was entered in one country, what users from other counties should see? 
  - using "start of the week" of country where this object is located 
  - using "start of the week" of country from their preferences

See also "The ISO 8601 definition for **week 01 is the week with the year's first Thursday in it**." - https://en.wikipedia.org/wiki/ISO_week_date#First_week

**Comment by ypid** [comment]
You are right. Without having more context. I would interpret this as something like [`Sep 15-Nov 15 Mo[-1],Mo[-1] +1 day,Mo[-1] +2 days,Mo[-1] +3 days,Mo[-1] +4 days,Mo[-1] +5 days 00:00-24:00`](http://openingh.openstreetmap.de/evaluation_tool/?EXP=PH%20off%3B%20Sep%20Mo[-1]%2CMo[-1]%20%2B1%20day%2CMo[-1]%20%2B2%20days%2CMo[-1]%20%2B3%20days%2CMo[-1]%20%2B4%20days%2CMo[-1]%20%2B5%20days%2000%3A00-24%3A00&DATE=1474994160000&lat=48.7769&lon=9.1844&mode=0) (+ the `+6 days … +7 days`). But the implementation for this is somehow broken. Also I limited the day offset to 6 because I did not think about such a use case yet.

**Comment by d1g** [comment]
@ypid, you're talking about last complete week that is created by wrapping into next month (case 2).

I was able to get definition "[the last full week in september from Monday through Saturday](http://www.openstreetmap.org/changeset/8655029)" (case 3).

Case 3 could be implemented using naive loop:

```
set nth_counter to 0;
Start with first day of the month, iterate +1 day until "calendar month" will change
if (current weekday of iteration is not equal to "start of the week") skip day(s) of incomplete week
if (current weekday of iteration is equal to "last day of the week") 
increment nth_counter and/or compute using nth_counter or store a nth recurring rule;

Note that "last full week in " and "first full week in " will resolve to 1 and 3-4 nth week, independently per each month+year.
```

Case 3 would result in 3-4 weeks, but I never seen any schedule "last full week in september from Monday through Saturday" Simpler "last Monday of the month" are far more common.
